Beginnings of a better celery UP metric

dannyroberts opened the pull request:
This extends the heartbeat concept to all queues, and drips a task into each every 10 seconds. This provides a reasonable measure of blockage, which I think should be the metric we use to determine whether celery is "UP" in the sense of whether there's service interruption. Each queue has its own threshold for what counts as DOWN-worthy blockage. (This should eventually be per-env, and be able to recognize when an env doesn't have/need a queue.)

I think this is pretty good for the binary UP/DOWN metric. For measuring celery "performance" while it's UP, I think I'll be relying more on the time-to-start metric. Dripping tasks into every queue also makes the time to start metric much less spotty.

dannyroberts committed:
Drip tasks into each queue to measure blockage

this basically expands the heartbeat concept to all queues

dannyroberts committed:
Use new all-queue heartbeats in heartbeat check

dannyroberts committed:
Respond to stickler

snopoke reviewed (commented):
I like the idea. One issue I see is that for queues that have very long expected blockage thresholds this will fill up the queue with a whole lot of heartbeat tasks. I guess that's not a big issue since they will get run very quickly but just wondering if we should try and prevent that in any way. It would require knowing there is at least one heartbeat task pending in the queue which maybe isn't that easy to do.

proteusvacuum commented on corehq/celery_monitoring/tasks.py:
```diff
@@ -0,0 +1,11 @@
+# Create one periodic_task named heartbeat__{queue} for each queue
+for queue, time_to_start_alert_threshold in settings.CELERY_HEARTBEAT_THRESHOLDS.items():
+    heartbeat = Heartbeat(queue)
+    locals()[heartbeat.periodic_task_name] = Heartbeat(queue).periodic_task
```
nit, I think you could use `heartbeat.periodic_task` here

dannyroberts replied:
You mean using `heartbeat.periodic_task.__name__` instead of `heartbeat.periodic_task_name`?

proteusvacuum replied:
no, I was just wondering if it was necessary to make another instance of the `Heartbeat` class here.

dannyroberts replied:
Oh! Thanks, totally missed that. Will fix

proteusvacuum reviewed (approved):
If you didn't want to have a whole bunch of tasks that pile up in the queue, you could do something similar to how we only create a single `AsyncRestore` task per user: 
https://github.com/dimagi/commcare-hq/blob/master/corehq/ex-submodules/casexml/apps/phone/restore.py#L630

dannyroberts commented:
@snopoke Yeah that's true. A sort of assumption of the current design is that the task itself runs very quickly. If a queue has been blocked up for 6 hours, the longest threshold we use, at a rate of 1 task every 10 seconds, that'll amount to 2160 heartbeat tasks in the queue. I think making 2.2k requests to redis and datadog should happen pretty fast, and the overhead of processing tasks through celery I think is also not that large. So basically, I think I'm cool with this even in that situation.

dannyroberts commented:
One annoying thing about testing this is that staging is set up to run without celery beat. I might be interested in revisiting the decision to set things up that way on staging. I think originally it was because we copied real data to staging and we didn't want scheduled SMS's to get sent, but maybe there's a more targeted way of preventing SMS's from getting sent that would allow us to test things that rely on celery beat more reliably.

dannyroberts committed:
Clean up duplicate Heartbeat instaniation in celery_monitoring tasks.py

millerdev commented on corehq/celery_monitoring/heartbeat.py:
```diff
@@ -0,0 +1,67 @@
+        queue = self.queue
+
+        def heartbeat():
+            hb = Heartbeat(queue)
```
Is it necessary to create a new instance here? Could it just use `self` instead?

dannyroberts replied:
Yeah it may not be necessary. I wanted to avoid having this task function be a closure (or rather rely on the context it gets from being a closure), just because that felt weird to me. I think given how this gets instantiated on the celery worker etc that it would actually work just fine like you're saying, but it felt a little harder to reason with to me for some reason so I did i this way

millerdev replied:
> I wanted to avoid having this task function be a closure (or rather rely on the context it gets from being a closure)...

Isn't it doing exactly that by using `queue` on this line?

Aside: the thing that makes this class hard to reason about for me is the non-standard use of "caching" (which doesn't really feel like caching in this context), as @emord pointed out.

dannyroberts replied:
Yeah you're right I think while in my mind that was somehow less entangled than reusing self, it maybe isn't given the way closures work. I can change this to use `self`. I will probably rename it ouside the closure (`hb = self`) just because `self` has a different conventional meaning inside a celery task, but see what you're saying that there's no need to create a new `Heartbeat` instance. I'll implement this.

dannyroberts replied:
Eh, screw it I'll just use self. I think what I was suggesting causes more confusion than it avoids. I can worry about it if we ever need to introduce a celery `self` arg.

emord commented on corehq/celery_monitoring/heartbeat.py:
```diff
@@ -0,0 +1,67 @@
+
+    @quickcache(['self.queue'], timeout=HEARTBEAT_CACHE_TIMEOUT.total_seconds())
+    def get_last_seen(self):
+        # This function relies on the cache getting set manually in mark_seen
```
I think this would be clearer with a django model instead of relying on quickcache. It'll be at most 20(?) records so it'll likely be cached in memory anyways.

Could also add a django admin model if desired

dannyroberts replied:
That's interesting. This felt simpler to me for some reason. Is there a particular failure mode you're thinking of with keeping it in redis, or a particular practical advantage for keeping it in postgres? Or were you more thinking about just how this looks in code? For now I'm inclined to keep it the way it is, but I'm interested in hearing more

emord replied:
A) I don't think we should use redis as a primary data store especially considering [cache eviction we've seen on prod](https://dimagi-dev.atlassian.net/browse/PLAT-136). (though debatable if redis vs datadog is our primary for this) 

B) Using a django model doesn't require explicit exception handling (could use get_or_create and create with a null last seen)

C) is datadog down or experiencing slow ingest rates? look at a django admin page for latest status

D) a django model more clearly documents the data we are capturing. Right now I think its kind of unclear what we're tracking

> This felt simpler to me for some reason

I think this is simpler if you are familiar with quickcache and can quickly get up to speed on this module (the comments help a lot). A django model may be more approachable to someone who doesn't have as good of an understanding of our custom caching. 

It also doesn't fit in with my mental model of what a "cache" is or is supposed to do so it feels like we're making quickcache much more generic than a "cache"

dannyroberts replied:
I'm going to roll this out as is, using redis as the backend here. I'm onboard that the code around this is confusing for someone who's not familiar with the internals of quickcache, and isn't a normal use-case for quickcache, so it'd probably be cleaner to use actual cache.set and cache.get statements.

I don't think I share the view that storing in postgres is meaningfully better. For comparison, the current heartbeat check does almost exactly what I'm doing here, storing in redis, and I don't think we've ever had an issue with that getting evicted.

That said, I'm open to revisiting and revising. I mostly would like to get a V1 of this out today so we can see how it works as a celery metric—we currently don't have a meaningful celery check.

dannyroberts replied:
And I do appreciate the reasoning and thoughtful review. (I probably should have led with that.) I'm going to make the changes to make the redis code more conventional and readable, and I'm sure I will refer back to the details in your last comment when iterating on this

emord replied:
To be fair, my main issue with using redis as the storage medium is mostly idealistic/expectation setting. Its way easier to be able to answer the question "what data do we collect and store on HQ?" by saying "look at postgres/models.py" as that sets an obvious place to look for the data we are tracking when approaching a new app. I don't think postgres is meaningfully better than redis here either. I do think that using django's API and conventions is meaningfully better than rolling our own or depending on some of our own custom code.

Anyways I think you addressed the main concern here as the code is clearer thanks to using cache directly instead of relying on quickcache's internals. And at least there's a class that defines the interface to find heartbeats.

My nits aren't really enough to block this PR. Copule of thoughts around this(wouldn't be an issue with django model):

1. Currently this manually sets a redis key `heartbeat:{queue_name}`. I don't foresee name conflicts coming up soon, but how do we define what these keys should be? Do we have an exhaustive list of the namespaces that are currently taken?
2. Would it make sense to have `HeartbeatCache` in the models.py file since its essentially a database model?

dannyroberts committed:
Rename method to Heartbeat.make_periodic_task to not hide side effect

dannyroberts committed:
Use new Heartbeat setup in check_celery instead of check_heartbeat

the celery check is the one that has currently been giving us problems
and that we want to replace with something better, which I think is this.
I think eventually we plan to remove the separate heartbeat check.

dannyroberts commented:
I conferred with Nick and Farid who I've been working with on recent celery issues, and we agreed it made sense to use this in place of the current celery check (instead of in place of the current heartbeat check). The last commit I added makes that switch. Eventually, I think we'll just remove the current heartbeat check.

esoergel commented on corehq/celery_monitoring/heartbeat.py:
```diff
@@ -0,0 +1,66 @@
+    def periodic_task_name(self):
+        return 'heartbeat__{}'.format(self.queue)
+
+    def make_periodic_task(self):
```
Might be worth noting in a docstring or comment that this method gets executed at startup, so we should take care that it doesn't make network calls or do too much work.

dannyroberts replied:
OK great, thanks, added

esoergel reviewed (approved):
Looks good.  I also support dropping the old heartbeat, since it does essentially just a subset of exactly this.  Are you waiting to gain confidence in the new implementation?

dannyroberts committed:
Merge branch 'master' into dmr/celery-monitoring

dannyroberts committed:
Clean up Heartbeat.make_periodic_task and add docstring

dannyroberts committed:
Make Heartbeat's use of redis more conventional and explicit

dannyroberts commented:
> Are you waiting to gain confidence in the new implementation?
I just had a few open ends from the PR discussion that I wanted to address, but which now I think have. Now just waiting for travis's blessing

dannyroberts committed:
Merge branch 'master' into dmr/celery-monitoring

dannyroberts committed:
Merge branch 'master' into dmr/celery-monitoring

dannyroberts committed:
Fix celery_monitoring tests

dannyroberts commented:
Woohoo!

dannyroberts commented on corehq/celery_monitoring/heartbeat.py:
```diff
@@ -0,0 +1,93 @@
+                    'commcare.celery.heartbeat.blockage_duration',
+                    self.get_blockage_duration(),
+                    tags=['celery_queue:{}'.format(self.queue)]
+                )
```
After deploying to india and production, everything looked great except one thing that was kind of flummoxing me: this new metric wasn't showing up in datadog at https://app.datadoghq.com/metric/summary?filter=commcare.celery.heartbeat.

After poking into it for a bit, I was fairly certain it's because (as I realized) `self.get_blockage_duration()` was a timedelta rather than a number of seconds. To verify I ran this block from the shell with `.total_seconds()` added, and then the metric showed up a couple minutes later!

Will add a fix to this to this followup PR: https://github.com/dimagi/commcare-hq/pull/23913.
